File: apps/web/src/lib/server/rate-limit.ts
/**
 * API rate limiting and usage tracking
 *
 * Hardcoded rate limits enforced per-instance (not per-user).
 * In an instance-per-user deployment model, each user gets their own database,
 * so these limits apply to each user's instance independently.
 */

import { getPool } from './db';

// ============================================================================
// Auth Rate Limiting (in-memory, for login endpoint)
// ============================================================================

interface AuthRateLimitRecord {
	count: number;
	resetAt: number;
}

const authAttempts = new Map<string, AuthRateLimitRecord>();

const AUTH_WINDOW_MS = 15 * 60 * 1000; // 15 minutes
const AUTH_MAX_ATTEMPTS = 5;

/**
 * Check if an IP is rate limited for auth/login attempts
 *
 * Uses in-memory storage (no DB required) with silent failure
 * to prevent enumeration attacks.
 *
 * @param ip - The IP address to check
 * @returns Object with `allowed` boolean and optional `retryAfter` in seconds
 */
export function checkAuthRateLimit(ip: string): { allowed: boolean; retryAfter?: number } {
	const now = Date.now();
	const record = authAttempts.get(ip);

	// Clean up expired records periodically (1% chance per request)
	if (Math.random() < 0.01) {
		for (const [key, value] of authAttempts) {
			if (value.resetAt < now) {
				authAttempts.delete(key);
			}
		}
	}

	// No record or expired - allow and start new window
	if (!record || record.resetAt < now) {
		authAttempts.set(ip, { count: 1, resetAt: now + AUTH_WINDOW_MS });
		return { allowed: true };
	}

	// Check if limit exceeded
	if (record.count >= AUTH_MAX_ATTEMPTS) {
		const retryAfter = Math.ceil((record.resetAt - now) / 1000);
		return { allowed: false, retryAfter };
	}

	// Increment count and allow
	record.count++;
	return { allowed: true };
}

/**
 * Get current auth rate limit status for an IP (for debugging/admin)
 */
export function getAuthRateLimitStatus(ip: string): {
	attempts: number;
	remaining: number;
	resetIn?: number;
} {
	const now = Date.now();
	const record = authAttempts.get(ip);

	if (!record || record.resetAt < now) {
		return { attempts: 0, remaining: AUTH_MAX_ATTEMPTS };
	}

	return {
		attempts: record.count,
		remaining: Math.max(0, AUTH_MAX_ATTEMPTS - record.count),
		resetIn: Math.ceil((record.resetAt - now) / 1000)
	};
}

// ============================================================================
// API Rate Limiting (database-backed, for chat/API endpoints)
// ============================================================================
import { env } from '$env/dynamic/private';
import type { Pool } from 'pg';

export class RateLimitError extends Error {
	constructor(
		message: string,
		public readonly current: number,
		public readonly limit: number,
		public readonly resetAt?: Date
	) {
		super(message);
		this.name = 'RateLimitError';
	}
}

/**
 * Rate limits applied to all instances
 * Configurable via environment variables with safe defaults
 */
export interface RateLimits {
	chatRequestsPerDay: number;
	chatTokensPerDay: number;
	backgroundJobsPerDay: number;
}

/**
 * Parse environment variable with validation
 * @param key Environment variable name
 * @param defaultValue Default value if not set
 * @param min Minimum allowed value (enforces safety)
 */
function parseEnvInt(key: string, defaultValue: number, min: number): number {
	const value = env[key];
	if (!value) return defaultValue;

	const parsed = parseInt(value, 10);
	if (Number.isNaN(parsed)) {
		console.warn(`Invalid value for ${key}: "${value}", using default ${defaultValue}`);
		return defaultValue;
	}

	// Enforce minimum to prevent disabling limits
	if (parsed < min) {
		console.warn(
			`Value for ${key} (${parsed}) is below minimum (${min}), using minimum instead`
		);
		return min;
	}

	return parsed;
}

/**
 * Monthly token limits based on TIER env var
 * - starter: 4M tokens/month (~$6 budget)
 * - pro: 15M tokens/month (~$20 budget)
 */
function getMonthlyTokenLimit(): number {
	const tier = env.TIER || 'starter';
	return tier === 'pro' ? 15_000_000 : 4_000_000;
}

export const DEFAULT_RATE_LIMITS: RateLimits = {
	chatRequestsPerDay: parseEnvInt('RATE_LIMIT_CHAT_DAILY', 1000, 1),
	chatTokensPerDay: parseEnvInt('RATE_LIMIT_TOKENS_DAILY', 500_000, 1000),
	backgroundJobsPerDay: parseEnvInt('RATE_LIMIT_JOBS_DAILY', 100, 1)
};

export interface TokenUsage {
	input: number;
	output: number;
	model: string;
}

/**
 * Rate limit information for HTTP headers
 */
export interface RateLimitInfo {
	limit: number;
	remaining: number;
	reset: number; // Unix timestamp
	retryAfter?: number; // Seconds until reset (only on limit exceeded)
}

/**
 * Get rate limit headers for HTTP response
 * Call this after checkRateLimit to add standard rate limit headers
 */
export function getRateLimitHeaders(info: RateLimitInfo): Record<string, string> {
	const headers: Record<string, string> = {
		'X-RateLimit-Limit': String(info.limit),
		'X-RateLimit-Remaining': String(Math.max(0, info.remaining)),
		'X-RateLimit-Reset': String(info.reset)
	};

	if (info.retryAfter !== undefined) {
		headers['Retry-After'] = String(info.retryAfter);
	}

	return headers;
}

/**
 * Check if an endpoint request would exceed rate limits
 *
 * SECURITY: Uses atomic check-then-increment to prevent race conditions.
 * All limits are checked in a single query BEFORE incrementing counts.
 * This prevents concurrent requests from bypassing limits.
 *
 * Uses server-side timestamps (UTC) to prevent client tampering.
 *
 * @param endpoint - The endpoint being called (e.g., 'chat', 'background_job')
 * @param estimatedTokens - Optional estimate of tokens this request will use (for pre-check)
 * @param limits - Rate limits to apply
 * @returns Rate limit info for response headers
 * @throws {RateLimitError} if limits would be exceeded
 */
export async function checkRateLimit(
	endpoint: string,
	estimatedTokens: number = 0,
	limits: RateLimits = DEFAULT_RATE_LIMITS
): Promise<RateLimitInfo> {
	const pool = getPool();
	const now = new Date();
	const dayBucket = getDayBucket(now);
	const monthBucket = getMonthBucket(now);

	// Get the limit for this endpoint
	const dailyRequestLimit =
		endpoint === 'chat'
			? limits.chatRequestsPerDay
			: endpoint === 'background_job'
				? limits.backgroundJobsPerDay
				: Number.MAX_SAFE_INTEGER;

	const monthlyLimit = getMonthlyTokenLimit();

	// ATOMIC: Check ALL limits in a single query BEFORE incrementing
	// This prevents race conditions where concurrent requests bypass limits
	const checkResult = await pool.query<{
		daily_requests: number;
		daily_tokens: number;
		monthly_tokens: number;
		status: 'ok' | 'daily_request_limit' | 'daily_token_limit' | 'monthly_token_limit';
	}>(
		`
		WITH current_usage AS (
			SELECT
				COALESCE((
					SELECT request_count FROM app.api_usage
					WHERE endpoint = $1 AND day_bucket = $2
				), 0) as daily_requests,
				COALESCE((
					SELECT SUM(token_count) FROM app.api_usage
					WHERE day_bucket = $2
				), 0) as daily_tokens,
				COALESCE((
					SELECT tokens_used FROM app.llm_usage
					WHERE month = $3
				), 0) as monthly_tokens
		)
		SELECT
			daily_requests,
			daily_tokens::bigint,
			monthly_tokens::bigint,
			CASE
				WHEN daily_requests >= $4 THEN 'daily_request_limit'
				WHEN daily_tokens + $7 > $5 THEN 'daily_token_limit'
				WHEN monthly_tokens + $7 > $6 THEN 'monthly_token_limit'
				ELSE 'ok'
			END as status
		FROM current_usage
		`,
		[
			endpoint,
			dayBucket,
			monthBucket,
			dailyRequestLimit,
			limits.chatTokensPerDay,
			monthlyLimit,
			estimatedTokens
		]
	);

	const check = checkResult.rows[0];

	// Throw appropriate error if any limit is exceeded
	if (check.status === 'daily_request_limit') {
		const resetAt = new Date(`${dayBucket}T00:00:00.000Z`);
		resetAt.setUTCDate(resetAt.getUTCDate() + 1);
		throw new RateLimitError(
			`Daily rate limit exceeded: ${check.daily_requests}/${dailyRequestLimit} requests`,
			check.daily_requests,
			dailyRequestLimit,
			resetAt
		);
	}

	if (check.status === 'daily_token_limit') {
		const resetAt = new Date(`${dayBucket}T00:00:00.000Z`);
		resetAt.setUTCDate(resetAt.getUTCDate() + 1);
		throw new RateLimitError(
			`Daily token limit exceeded: ${check.daily_tokens}/${limits.chatTokensPerDay} tokens`,
			Number(check.daily_tokens),
			limits.chatTokensPerDay,
			resetAt
		);
	}

	if (check.status === 'monthly_token_limit') {
		const resetAt = new Date(`${monthBucket}T00:00:00.000Z`);
		resetAt.setUTCMonth(resetAt.getUTCMonth() + 1);
		const tier = env.TIER || 'starter';
		throw new RateLimitError(
			`Monthly token limit exceeded (${tier} tier): ${check.monthly_tokens}/${monthlyLimit} tokens`,
			Number(check.monthly_tokens),
			monthlyLimit,
			resetAt
		);
	}

	// Only NOW increment request count (after all checks pass)
	await pool.query(
		`
		INSERT INTO app.api_usage (endpoint, day_bucket, request_count)
		VALUES ($1, $2, 1)
		ON CONFLICT (endpoint, day_bucket)
		DO UPDATE SET
			request_count = app.api_usage.request_count + 1,
			updated_at = NOW()
		`,
		[endpoint, dayBucket]
	);

	// Calculate reset time (next day at midnight UTC)
	const resetAt = new Date(`${dayBucket}T00:00:00.000Z`);
	resetAt.setUTCDate(resetAt.getUTCDate() + 1);

	// Return rate limit info for response headers
	return {
		limit: dailyRequestLimit,
		remaining: dailyRequestLimit - check.daily_requests - 1, // -1 for this request
		reset: Math.floor(resetAt.getTime() / 1000)
	};
}

/**
 * Record token usage after a successful API call
 *
 * IMPORTANT: This function MUST be called after checkRateLimit() to ensure
 * the api_usage row exists. Calling this without checkRateLimit() will throw an error.
 *
 * Note: Request count is already incremented by checkRateLimit().
 * This function only updates token counts and cost estimates.
 *
 * Also records to LLM metering tables for multi-tenant billing:
 * - app.llm_usage: monthly aggregates
 * - app.llm_requests: individual request log
 *
 * @throws {Error} if no row exists (checkRateLimit not called first)
 * @throws {Error} if token counts exceed safe integer range
 */
export async function recordUsage(endpoint: string, tokens: TokenUsage): Promise<void> {
	const pool = getPool();
	const now = new Date();
	const dayBucket = getDayBucket(now);
	const monthBucket = getMonthBucket(now);
	const cost = calculateCost(tokens);
	const costCents = Math.round(cost * 100);

	// Validate token counts don't exceed safe integer range
	// PostgreSQL INTEGER max is 2,147,483,647
	// JavaScript MAX_SAFE_INTEGER is 9,007,199,254,740,991
	const totalTokens = tokens.input + tokens.output;
	if (
		totalTokens > Number.MAX_SAFE_INTEGER ||
		totalTokens > 2147483647 ||
		tokens.input > 2147483647 ||
		tokens.output > 2147483647
	) {
		throw new Error(
			`Token count exceeds safe range: input=${tokens.input}, output=${tokens.output}, total=${totalTokens}`
		);
	}

	// Update daily usage (request count already incremented by checkRateLimit)
	const result = await pool.query(
		`
		UPDATE app.api_usage
		SET
			token_count = token_count + $3,
			input_tokens = input_tokens + $4,
			output_tokens = output_tokens + $5,
			estimated_cost_usd = estimated_cost_usd + $6,
			updated_at = NOW()
		WHERE endpoint = $1 AND day_bucket = $2
		`,
		[endpoint, dayBucket, totalTokens, tokens.input, tokens.output, cost]
	);

	// Verify the UPDATE affected at least one row
	// If no rows were affected, it means checkRateLimit() wasn't called first
	if (result.rowCount === 0) {
		throw new Error(
			`Failed to record usage: no row found for endpoint="${endpoint}", day="${dayBucket}". ` +
				`Ensure checkRateLimit() is called before recordUsage().`
		);
	}

	// Record to LLM metering tables for multi-tenant billing
	// These are used by Atlas for billing calculations
	await Promise.all([
		// Update monthly aggregate
		pool.query(
			`
			INSERT INTO app.llm_usage (month, tokens_used, cost_cents)
			VALUES ($1, $2, $3)
			ON CONFLICT (month)
			DO UPDATE SET
				tokens_used = app.llm_usage.tokens_used + $2,
				cost_cents = app.llm_usage.cost_cents + $3
			`,
			[monthBucket, totalTokens, costCents]
		),
		// Log individual request
		pool.query(
			`
			INSERT INTO app.llm_requests (model, input_tokens, output_tokens)
			VALUES ($1, $2, $3)
			`,
			[tokens.model, tokens.input, tokens.output]
		)
	]);
}

/**
 * Get current usage statistics for display in UI
 */
export interface UsageStats {
	dailyRequests: number;
	dailyTokens: number;
	dailyCost: number;
	limits: RateLimits;
}

export async function getUsageStats(endpoint: string): Promise<UsageStats> {
	const pool = getPool();
	const now = new Date();
	const dayBucket = getDayBucket(now);
	const limits = DEFAULT_RATE_LIMITS;

	const [dailyUsage, totalDailyTokens, dailyCost] = await Promise.all([
		getDailyUsage(pool, endpoint, dayBucket),
		getTotalDailyTokens(pool, dayBucket),
		getDailyCost(pool, dayBucket)
	]);

	return {
		dailyRequests: dailyUsage.requestCount,
		dailyTokens: totalDailyTokens,
		dailyCost,
		limits
	};
}

// ============================================================================
// Helper Functions
// ============================================================================

function getDayBucket(date: Date): string {
	const d = new Date(date);
	d.setUTCHours(0, 0, 0, 0);
	// Return date as string in YYYY-MM-DD format to match PostgreSQL DATE type
	// This matches Rust's NaiveDate representation
	return d.toISOString().split('T')[0];
}

function getMonthBucket(date: Date): string {
	const d = new Date(date);
	d.setUTCHours(0, 0, 0, 0);
	d.setUTCDate(1); // First day of month
	// Return date as string in YYYY-MM-DD format (first of month)
	return d.toISOString().split('T')[0];
}

interface DailyUsage {
	requestCount: number;
	tokenCount: number;
}

async function getDailyUsage(pool: Pool, endpoint: string, dayBucket: string): Promise<DailyUsage> {
	const result = await pool.query(
		`
		SELECT
			COALESCE(request_count, 0) as request_count,
			COALESCE(token_count, 0) as token_count
		FROM app.api_usage
		WHERE endpoint = $1 AND day_bucket = $2
		`,
		[endpoint, dayBucket]
	);

	return {
		requestCount: result.rows[0]?.request_count ?? 0,
		tokenCount: result.rows[0]?.token_count ?? 0
	};
}

async function getTotalDailyTokens(pool: Pool, dayBucket: string): Promise<number> {
	const result = await pool.query(
		`
		SELECT COALESCE(SUM(token_count), 0)::int as total_tokens
		FROM app.api_usage
		WHERE day_bucket = $1
		`,
		[dayBucket]
	);

	return result.rows[0]?.total_tokens ?? 0;
}

async function getDailyCost(pool: Pool, dayBucket: string): Promise<number> {
	const result = await pool.query(
		`
		SELECT COALESCE(SUM(estimated_cost_usd), 0)::numeric as total_cost
		FROM app.api_usage
		WHERE day_bucket = $1
		`,
		[dayBucket]
	);

	return parseFloat(result.rows[0]?.total_cost ?? 0);
}

/**
 * Calculate estimated cost based on token usage and model
 * Pricing as of November 2024 (subject to change)
 */
function calculateCost(tokens: TokenUsage): number {
	const modelLower = tokens.model.toLowerCase();

	// Anthropic Claude pricing (per million tokens)
	let inputPrice = 3.0; // Default to Sonnet
	let outputPrice = 15.0;

	if (modelLower.includes('sonnet')) {
		if (modelLower.includes('4')) {
			inputPrice = 3.0;
			outputPrice = 15.0;
		} else {
			inputPrice = 3.0;
			outputPrice = 15.0;
		}
	} else if (modelLower.includes('haiku')) {
		inputPrice = 0.8;
		outputPrice = 4.0;
	} else if (modelLower.includes('opus')) {
		inputPrice = 15.0;
		outputPrice = 75.0;
	} else if (modelLower.includes('gpt')) {
		// OpenAI pricing estimates
		if (modelLower.includes('gpt-5')) {
			inputPrice = 5.0;
			outputPrice = 15.0;
		} else if (modelLower.includes('oss')) {
			inputPrice = 0.0; // Open source models via AI Gateway
			outputPrice = 0.0;
		} else {
			inputPrice = 2.5;
			outputPrice = 10.0;
		}
	} else if (modelLower.includes('gemini')) {
		inputPrice = 1.25;
		outputPrice = 5.0;
	}

	const inputCost = (tokens.input * inputPrice) / 1_000_000;
	const outputCost = (tokens.output * outputPrice) / 1_000_000;

	return inputCost + outputCost;
}

File: core/src/api/rate_limit.rs
//! Rate limiting and usage tracking for API endpoints
//!
//! Provides hardcoded rate limits that are enforced per-instance (not per-user).
//! In an instance-per-user deployment model, each user gets their own database,
//! so these limits apply to each user's instance independently.

use chrono::{DateTime, NaiveDate, Utc};
use sqlx::SqlitePool;
use thiserror::Error;
use uuid::Uuid;

#[derive(Error, Debug)]
pub enum RateLimitError {
    #[error("Daily rate limit exceeded: {current}/{limit} requests. Resets at {reset_at}")]
    DailyLimitExceeded {
        current: i32,
        limit: i32,
        reset_at: String,
    },

    #[error("Daily token limit exceeded: {current}/{limit} tokens. Resets at {reset_at}")]
    TokenLimitExceeded {
        current: i32,
        limit: i32,
        reset_at: String,
    },

    #[error("Database error: {0}")]
    DatabaseError(#[from] sqlx::Error),
}

/// Rate limits applied to all instances
/// In instance-per-user model, each user gets these same limits
/// Configurable via environment variables with safe defaults
#[derive(Debug, Clone)]
pub struct RateLimits {
    /// Maximum chat requests per day
    pub chat_requests_per_day: i32,
    /// Maximum tokens (input + output) per day across all endpoints
    pub chat_tokens_per_day: i32,
    /// Maximum background job LLM calls per day
    pub background_jobs_per_day: i32,
}

impl Default for RateLimits {
    fn default() -> Self {
        Self {
            chat_requests_per_day: 1000,  // 1000 chat requests/day
            chat_tokens_per_day: 500_000, // 500K tokens/day (~$1.50/day worst case)
            background_jobs_per_day: 100, // 100 background LLM jobs/day
        }
    }
}

impl RateLimits {
    /// Load rate limits from environment variables with safe defaults
    ///
    /// Environment variables:
    /// - RATE_LIMIT_CHAT_DAILY (default: 1000, min: 1)
    /// - RATE_LIMIT_TOKENS_DAILY (default: 500000, min: 1000)
    /// - RATE_LIMIT_JOBS_DAILY (default: 100, min: 1)
    pub fn from_env() -> Self {
        fn parse_env(key: &str, default: i32, min: i32) -> i32 {
            std::env::var(key)
                .ok()
                .and_then(|s| s.parse().ok())
                .unwrap_or(default)
                .max(min) // Enforce minimum to prevent disabling limits
        }

        Self {
            chat_requests_per_day: parse_env("RATE_LIMIT_CHAT_DAILY", 1000, 1),
            chat_tokens_per_day: parse_env("RATE_LIMIT_TOKENS_DAILY", 500_000, 1000),
            background_jobs_per_day: parse_env("RATE_LIMIT_JOBS_DAILY", 100, 1),
        }
    }

    /// Create custom rate limits (useful for testing or custom tiers)
    pub fn custom(chat_per_day: i32, tokens_per_day: i32, jobs_per_day: i32) -> Self {
        Self {
            chat_requests_per_day: chat_per_day,
            chat_tokens_per_day: tokens_per_day,
            background_jobs_per_day: jobs_per_day,
        }
    }
}

/// Token usage information
#[derive(Debug, Clone)]
pub struct TokenUsage {
    pub input: u32,
    pub output: u32,
    pub model: String,
}

impl TokenUsage {
    pub fn new(input: u32, output: u32, model: impl Into<String>) -> Self {
        Self {
            input,
            output,
            model: model.into(),
        }
    }

    pub fn total(&self) -> u32 {
        self.input + self.output
    }
}

/// Check if an endpoint request would exceed rate limits
///
/// Atomically increments the request count and then checks if limits are exceeded.
/// Uses optimistic locking: if limit is exceeded after increment, the request has
/// already been counted. This is acceptable for daily limits as it only allows
/// one extra request rather than unbounded concurrent bypass.
///
/// Uses server-side timestamps (UTC) to prevent client tampering.
///
/// # Arguments
/// * `pool` - Database connection pool
/// * `endpoint` - Endpoint identifier (e.g., "chat", "background_job")
/// * `limits` - Rate limit configuration
///
/// # Returns
/// * `Ok(())` if within limits
/// * `Err(RateLimitError)` if limits would be exceeded
pub async fn check_rate_limit(
    pool: &SqlitePool,
    endpoint: &str,
    limits: &RateLimits,
) -> Result<(), RateLimitError> {
    let now = Utc::now();
    let day_bucket = get_day_bucket(now);

    // Get the limit for this endpoint
    let daily_request_limit = match endpoint {
        "chat" => limits.chat_requests_per_day,
        "background_job" => limits.background_jobs_per_day,
        _ => i32::MAX, // No limit for other endpoints
    };

    // Atomically increment request count and get new total
    // This prevents race conditions by using database-level atomicity
    // SQLite uses 'excluded' for conflict values, and we need to reference the table directly
    let usage_id = Uuid::new_v4().to_string();
    let new_count = sqlx::query!(
        r#"
        INSERT INTO app_api_usage
            (id, endpoint, day_bucket, request_count)
        VALUES ($1, $2, $3, 1)
        ON CONFLICT (endpoint, day_bucket)
        DO UPDATE SET
            request_count = request_count + 1,
            updated_at = datetime('now')
        RETURNING request_count as "count!: i32"
        "#,
        usage_id,
        endpoint,
        day_bucket
    )
    .fetch_one(pool)
    .await?
    .count;

    // Check if we've exceeded the limit (after incrementing)
    if new_count > daily_request_limit {
        // Calculate reset time (next day at midnight UTC)
        let reset_at = day_bucket
            .succ_opt()
            .unwrap_or(day_bucket)
            .and_hms_opt(0, 0, 0)
            .unwrap()
            .and_utc()
            .to_rfc3339();

        return Err(RateLimitError::DailyLimitExceeded {
            current: new_count,
            limit: daily_request_limit,
            reset_at,
        });
    }

    // Check daily token limit (applies to all endpoints combined)
    let total_daily_tokens = get_total_daily_tokens(pool, day_bucket).await?;
    if total_daily_tokens > limits.chat_tokens_per_day {
        // Calculate reset time (next day at midnight UTC)
        let reset_at = day_bucket
            .succ_opt()
            .unwrap_or(day_bucket)
            .and_hms_opt(0, 0, 0)
            .unwrap()
            .and_utc()
            .to_rfc3339();

        return Err(RateLimitError::TokenLimitExceeded {
            current: total_daily_tokens,
            limit: limits.chat_tokens_per_day,
            reset_at,
        });
    }

    Ok(())
}

/// Record token usage after a successful API call
///
/// **IMPORTANT**: This function MUST be called after `check_rate_limit()` to ensure
/// the `api_usage` row exists. Calling this without `check_rate_limit()` will return an error.
///
/// Note: Request count is already incremented by `check_rate_limit()`.
/// This function only updates token counts and cost estimates.
///
/// # Arguments
/// * `pool` - Database connection pool
/// * `endpoint` - Endpoint identifier
/// * `tokens` - Token usage information
///
/// # Returns
/// * `Ok(())` on success
/// * `Err(sqlx::Error::RowNotFound)` if no row exists (check_rate_limit not called first)
/// * `Err(sqlx::Error::Protocol)` if token counts exceed i32::MAX
/// * `Err(sqlx::Error)` on other database errors
pub async fn record_usage(
    pool: &SqlitePool,
    endpoint: &str,
    tokens: TokenUsage,
) -> Result<(), sqlx::Error> {
    let now = Utc::now();
    let day_bucket = get_day_bucket(now);

    let cost = calculate_cost(&tokens);

    // Convert token counts with overflow protection
    let total_tokens = i32::try_from(tokens.total())
        .map_err(|_| sqlx::Error::Protocol("Token count exceeds i32::MAX".into()))?;
    let input_tokens = i32::try_from(tokens.input)
        .map_err(|_| sqlx::Error::Protocol("Input token count exceeds i32::MAX".into()))?;
    let output_tokens = i32::try_from(tokens.output)
        .map_err(|_| sqlx::Error::Protocol("Output token count exceeds i32::MAX".into()))?;

    // Update token usage (request count already incremented by check_rate_limit)
    let result = sqlx::query!(
        r#"
        UPDATE app_api_usage
        SET
            token_count = token_count + $3,
            input_tokens = input_tokens + $4,
            output_tokens = output_tokens + $5,
            estimated_cost_usd = estimated_cost_usd + $6,
            updated_at = datetime('now')
        WHERE endpoint = $1 AND day_bucket = $2
        "#,
        endpoint,
        day_bucket,
        total_tokens,
        input_tokens,
        output_tokens,
        cost
    )
    .execute(pool)
    .await?;

    // Verify the UPDATE affected at least one row
    // If no rows were affected, it means check_rate_limit() wasn't called first
    if result.rows_affected() == 0 {
        return Err(sqlx::Error::RowNotFound);
    }

    Ok(())
}

/// Get current usage statistics for display in UI
#[derive(Debug, Clone)]
pub struct UsageStats {
    pub daily_requests: i32,
    pub daily_tokens: i32,
    pub daily_cost: f64,
    pub limits: RateLimits,
}

pub async fn get_usage_stats(pool: &SqlitePool, endpoint: &str) -> Result<UsageStats, sqlx::Error> {
    let now = Utc::now();
    let day_bucket = get_day_bucket(now);
    let limits = RateLimits::default();

    let daily_usage = get_daily_usage(pool, endpoint, day_bucket).await?;
    let total_daily_tokens = get_total_daily_tokens(pool, day_bucket).await?;
    let daily_cost = get_daily_cost(pool, day_bucket).await?;

    Ok(UsageStats {
        daily_requests: daily_usage.request_count,
        daily_tokens: total_daily_tokens,
        daily_cost,
        limits,
    })
}

// ============================================================================
// Helper Functions
// ============================================================================

fn get_day_bucket(dt: DateTime<Utc>) -> NaiveDate {
    dt.date_naive()
}

#[derive(Debug)]
struct DailyUsage {
    request_count: i32,
    #[allow(dead_code)]
    token_count: i32,
}

async fn get_daily_usage(
    pool: &SqlitePool,
    endpoint: &str,
    day_bucket: NaiveDate,
) -> Result<DailyUsage, sqlx::Error> {
    // SQLite needs explicit type annotations for COALESCE results
    let result = sqlx::query!(
        r#"
        SELECT
            COALESCE(request_count, 0) as "request_count!: i32",
            COALESCE(token_count, 0) as "token_count!: i32"
        FROM app_api_usage
        WHERE endpoint = $1 AND day_bucket = $2
        "#,
        endpoint,
        day_bucket
    )
    .fetch_optional(pool)
    .await?;

    Ok(result
        .map(|r| DailyUsage {
            request_count: r.request_count,
            token_count: r.token_count,
        })
        .unwrap_or(DailyUsage {
            request_count: 0,
            token_count: 0,
        }))
}

async fn get_total_daily_tokens(
    pool: &SqlitePool,
    day_bucket: NaiveDate,
) -> Result<i32, sqlx::Error> {
    let result = sqlx::query!(
        r#"
        SELECT COALESCE(SUM(token_count), 0) as "total_tokens!: i64"
        FROM app_api_usage
        WHERE day_bucket = $1
        "#,
        day_bucket
    )
    .fetch_one(pool)
    .await?;

    Ok(result.total_tokens as i32)
}

async fn get_daily_cost(pool: &SqlitePool, day_bucket: NaiveDate) -> Result<f64, sqlx::Error> {
    let result = sqlx::query!(
        r#"
        SELECT COALESCE(SUM(estimated_cost_usd), 0.0) as "total_cost!: f64"
        FROM app_api_usage
        WHERE day_bucket = $1
        "#,
        day_bucket
    )
    .fetch_one(pool)
    .await?;

    Ok(result.total_cost)
}

/// Calculate estimated cost based on token usage and model
/// Pricing as of November 2024 (subject to change)
fn calculate_cost(tokens: &TokenUsage) -> f64 {
    let model_lower = tokens.model.to_lowercase();

    // Anthropic Claude pricing (per million tokens)
    let (input_price, output_price): (f64, f64) = if model_lower.contains("sonnet") {
        if model_lower.contains("4") {
            (3.00, 15.00) // Claude Sonnet 4
        } else {
            (3.00, 15.00) // Claude 3.5 Sonnet
        }
    } else if model_lower.contains("haiku") {
        (0.80, 4.00) // Claude 3.5 Haiku
    } else if model_lower.contains("opus") {
        (15.00, 75.00) // Claude 3 Opus
    } else {
        // Default to Sonnet pricing if unknown
        (3.00, 15.00)
    };

    let input_cost = (tokens.input as f64) * input_price / 1_000_000.0;
    let output_cost = (tokens.output as f64) * output_price / 1_000_000.0;

    input_cost + output_cost
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_calculate_cost_sonnet() {
        let tokens = TokenUsage::new(1000, 500, "claude-sonnet-4");
        let cost = calculate_cost(&tokens);

        // 1000 * 0.003 / 1000 + 500 * 0.015 / 1000 = 0.003 + 0.0075 = 0.0105
        assert!(cost > 0.01);
        assert!(cost < 0.02);
    }

    #[test]
    fn test_calculate_cost_haiku() {
        let tokens = TokenUsage::new(10000, 5000, "claude-haiku-3.5");
        let cost = calculate_cost(&tokens);

        // Should be much cheaper than Sonnet
        assert!(cost < 0.05);
    }

    #[test]
    fn test_rate_limits_default() {
        let limits = RateLimits::default();
        assert_eq!(limits.chat_requests_per_day, 1000);
        assert_eq!(limits.chat_tokens_per_day, 500_000);
        assert_eq!(limits.background_jobs_per_day, 100);
    }

    #[test]
    fn test_token_usage_total() {
        let tokens = TokenUsage::new(1000, 500, "claude-sonnet-4");
        assert_eq!(tokens.total(), 1500);
    }
}
